### PR TITLE
DS3 Pad Handler

### PR DESCRIPTION
**Pull request by RipleyTom**
First the good news:
-It supports pressure buttons
-It supports sixaxis
-It supports force feedback

Bad news:
-Doesn't work with scptoolkit
-You need to install 'generic' driver like WinUSB(I recommend using Zadig: https://zadig.akeo.ie/ )
-Doesn't support bluetooth yet(I don't have any BT dongle to test it with being the main reason)

Note:
For the DS3 to start sending reports you need to press the PS button in the middle of the pad(it was the same on the PS3).

### CODE REVIEW

**Comment by yatsukiko** [comment]
Rpcs3 sees DS3 and it works in-game, though I can't load any saves, it freezes when it's supposed to show saves(rpcs3's save menu). I can load games normally if they have their own save loader. I also can't type using OSK nor physical keyboard. In nier main menu scrolling through options is too fast, even when limited to 30 fps.

File: rpcs3/ds3_pad_handler.cpp [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+		0xff, 0x27, 0x10, 0x00, 0x32,
+		0xff, 0x27, 0x10, 0x00, 0x32,
+		0x00, 0x00, 0x00, 0x00, 0x00
+	};
+
+	report_buf[2] = ds3dev->large_motor;
+	report_buf[4] = ds3dev->small_motor;
+
+	libusb_control_transfer(ds3dev->handle, LIBUSB_ENDPOINT_OUT | LIBUSB_REQUEST_TYPE_CLASS | LIBUSB_RECIPIENT_INTERFACE, 0x09, 0x0201, 0, report_buf, sizeof(report_buf), 0);
+}
```
Comment (kd-11) [inline rpcs3/ds3_pad_handler.cpp]: consider adding explicit enums or even a single macro to help with readability. I'm talking about the raw hex values 0x09 and 0x0201

**Comment by JarnoVgr** [comment]
The creator of ScpToolkit is actively working a new unified way to allow DS3 and DS4 support on the same Bluetooth dongle with generic drivers. It might be good to follow: https://forums.vigem.org/topic/242/bluetooth-filter-driver-for-ds3-compatibility-research-notes

File: rpcs3/ds3_pad_handler.cpp [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+	else
+		LOG_SUCCESS(HLE, "[DS3] Controllers found: %d", controllers.size());
+
+	is_init = true;
+	return true;
+}
+
+std::vector<std::string> ds3_pad_handler::ListDevices()
+{
+	std::vector<std::string> ds4_pads_list;
```
Comment (gidan80) [inline rpcs3/ds3_pad_handler.cpp]: Typo?

File: rpcs3/ds3_pad_handler.cpp [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+	{
+		m_dev = bindings[i].first;
+		auto thepad = bindings[i].second;
+		auto profile = m_dev->config;
+
+		thepad->m_port_status = CELL_PAD_STATUS_CONNECTED | CELL_PAD_STATUS_ASSIGN_CHANGES;
+
+		if (get_data(m_dev))
+		{
+			if (get_data(m_dev))
```
Comment (gidan80) [inline rpcs3/ds3_pad_handler.cpp]: Same if inside another if

**Comment by RipleyTom** [comment]
Added enums for all the values used by usb functions.
Fixed the issue with overlays.

**Comment by vsub** [comment]
One question,if this supports sixaxis,is it possible some kind of feature to be implemented on the XInput side?
Pressing a button to execute a movement like on Dolphin for example
This will allow games that require sixaxis to work properly on xinput gamepad or a keyboard

**Comment by RainbowCookie32** [comment]
> One question,if this supports sixaxis,is it possible some kind of feature to be implemented on the XInput side?
> Pressing a button to execute a movement like on Dolphin for example
> This will allow games that require sixaxis to work properly on xinput gamepad or a keyboard

This is support for the DS3, that already supports Sixaxis. XInput devices don't, so some kind of emulation would be needed, which is unrelated to this PR.

**Comment by yatsukiko** [comment]
Overlays works now, but there is still problem with nier(gestalt)'s any list scrolling too fast. I tried Persona 5, Diva F 2nd and Drakengard 3, there weren't any problems with them.

**Comment by Zangetsu38** [comment]
So Like i have said in tester, good works for me, but i have see problem on gamepad test wxxith value "G=6" com=pared DS4 and on PS3 that is broken

**Comment by Daltz333** [comment]
Indeed. A new standalone suite of drivers is being developed for DS3 USB and Bluetooth. These drivers will be WHQL certified. Follow progress at the ViGEm Discord at https://discord.gg/QTJpBX5

**Comment by Krusher97** [comment]
Tested Folklore,Ni no Kuni, The Godfather and Toy Story 3
Out of those 4 only Toy Story 3 accepted inputs. Folklore and Ni No Kuni accept no inputs from the DS3 while Godfather stops at the reconnect wireless sixaxis controller prompt.

**Comment by RipleyTom** [comment]
Added support for disconnecting/reconnecting pads(still no bt though) and fixed some bugs, hopefully should work with everything now.

**Comment by RipleyTom** [comment]
Just a note to say I previously recommended libusb-win32 drivers but it's probably better to go with WinUSB driver which is the official generic microsoft driver for usb and work as well.

**Comment by Dokman** [comment]
My DS3 with WinUSB does not do anything in any game

**Comment by RipleyTom** [comment]
Forgot to mention you need to press the PS button in the middle of the pad for it to start sending reports, afaik it's the case on ps3 too.
Otherwise make sure it's properly configured in pad config.
Also it doesn't support any knockoff atm.

**Comment by Dokman** [comment]
Kingdom Hearts 2.5 BLES02070 Working Well no crashes atm
Infamous Loading but not Ingame

File: rpcs3/ds3_pad_handler.cpp [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+std::shared_ptr<ds3_pad_handler::ds3_device> ds3_pad_handler::get_device(const std::string& padId)
+{
+	if (!Init())
+		return nullptr;
+
+	size_t pos = padId.find(m_name_string);
+	if (pos == std::string::npos)
+		return nullptr;
+
+	int pad_number = std::stoi(padId.substr(pos + 9));
```
Comment (kd-11) [inline rpcs3/ds3_pad_handler.cpp]: Could've just done (padId[pos + 9] - '0') assuming no more than 9 controllers available at any one time, but this is fine too.

**Comment by marcoluc97** [comment]
@RipleyTom i don't know if that can help this PR but i see an update from the creator of ScpToolkit, if you remember there was a WIP of a new unified way to allow DS3 and DS4 support on the same Bluetooth dongle with generic drivers. Now is released at this link: [https://forums.vigem.org/topic/362/playstation-r-3-peripherals-unofficial-bluetooth-drivers-bthps3](url)
I hope this if useful. Have a nice day.

**Pull request merged**